plugins/com.google.cloud.tools.eclipse.usagetracker/src/com/google/cloud/tools/eclipse/usagetracker/AnalyticsPingManager.java

```diff
@@ -0,0 +1,116 @@
+import java.util.Map;
+import java.util.UUID;
+
+/**
+ * Provides methods that report plugin-specific events to Analytics.
+ */
+public class AnalyticsPingManager {
+
+  private static final String ANALYTICS_COLLECTION_URL = "https://ssl.google-analytics.com/collect";
+  // TODO: retrieve tracking ID from a system environment variable.
+  private static final String ANALYTICS_TRACKING_ID = "@ga.tracking.id@";
+  private static final String APPLICATION_NAME = "gcloud-eclipse-tools";
+
+  // TODO: see if there a way to get a unique ID for each Eclipse installation.
+  private static final String CLIENT_ID = UUID.randomUUID().toString();
+
+  // Fixed-value query parameters present in every ping, and their fixed values:
+  //
+  // For the semantics of each parameter, consult the followings:
+  //
+  // https://github.com/google/cloud-reporting/blob/master/src/main/java/com/google/cloud/metrics/MetricsUtils.java#L183
+  // https://developers.google.com/analytics/devguides/collection/protocol/v1/reference
+  private static final ImmutableMap<String, String> STANDARD_PARAMETERS =
+      ImmutableMap.<String, String>builder()
+          .put("v", "1")  // Google Analytics Measurement Protocol version
+          .put("tid", ANALYTICS_TRACKING_ID) // tracking ID
+          .put("ni", "0")  // Non-interactive? Report as interactive.
+          .put("t", "pageview")  // Hit type
+          .put("cd21", "1")  // Yes, this ping is a virtual "page".
+          .put("cd16", "0")  // Internal user? No.
+          .put("cd17", "0")  // User signed in? We will ignore this.
+          .put("cid",  CLIENT_ID)
+          .build();
+
+  private static final PercentEscaper ESCAPER = new PercentEscaper(".-_", false);
+
+  private static boolean isTrackingIdDefined() {
+    return ANALYTICS_TRACKING_ID != null && ANALYTICS_TRACKING_ID.startsWith("UA-");
+  }
+
+  public static void sendPing(String eventType, String eventName,
+      String metadataKey, String metadataValue) {
+    Map<String, String> parametersMap = Maps.newHashMap(STANDARD_PARAMETERS);
+
+    parametersMap.put("cd19", eventType);
+    parametersMap.put("cd20", eventName);
+
+    String virtualPageUrl = "/virtual/" + APPLICATION_NAME + "/" + eventType + "/" + eventName;
+    parametersMap.put("dp", virtualPageUrl);
+
+    if (metadataKey != null) {
+      // Event metadata are passed as a (virtual) page title.
+      String virtualPageTitle =
+          metadataKey + "=" + (metadataValue != null ? metadataValue : "null");
+      parametersMap.put("dt", virtualPageTitle);
+    }
+
+    sendPostRequest(parametersMap);
+  }
+
+  private static void sendPostRequest(Map<String, String> parametersMap) {
+    if (Platform.inDevelopmentMode() || !isTrackingIdDefined()) {
+      return;
+    }
+
+    String parametersString = getParametersString(parametersMap);
+    try {
+      URL url = new URL(ANALYTICS_COLLECTION_URL);
+      HttpURLConnection connection = (HttpURLConnection) url.openConnection();
```

Comment:
I don't think any of your tests exercise the percent escaping functionality. Please add some that do.